src/net.cpp

```diff
@@ -2410,7 +2414,7 @@ bool CConnman::Bind(const CService &addr, unsigned int flags, NetPermissionFlags
         return false;
     }
 
-    if (addr.IsRoutable() && fDiscover && !(flags & BF_DONT_ADVERTISE) && !(permissions & PF_NOBAN)) {
+    if (addr.IsRoutable() && fDiscover && !(flags & BF_DONT_ADVERTISE) && !NetPermissions::HasAnyFlag(permissions, NetPermissionFlags::NoBan)) {
```

Comment:
Why not `NetPermissions::HasFlag` ?

Reply:
Hm, the tests pass with `HasFlag()` but ISTM that would represent a potential change in behavior in a refactoring pull, as `HasFlag()` returns `(flags & f) == f` instead of `flags & f` and `NetPermissionFlags::NoBan` is a multi-flag (`NoBan = (1U << 4) | Download`). LMK if I'm confused here though.

Aside, this line was last changed in commit bb145c90502:
```diff
-    if (addr.IsRoutable() && fDiscover && (permissions & PF_NOBAN) == 0) {
+    if (addr.IsRoutable() && fDiscover && !(flags & BF_DONT_ADVERTISE) && !(permissions & PF_NOBAN)) {
```
There do seem to be some oddities and unused code WRT the net permissions flags that could be cleaned up in a separate patch.

Reply:
You are right. Your change is equivalent to the current code. But this raises additional questions:

1. current tests are not reliable if they
> pass with `HasFlag()`

2. `HasAnyFlag(permissions, flag_a | flag_b)` is much easier to read and reason about than `HasAnyFlag(permissions, fancy_multiflag)`

3. in general, with named multiflags, negation of `HasFlag` could have ambiguous meaning. Does this code do what it is supposed to do: https://github.com/bitcoin/bitcoin/blob/590e49ccf2af27c6c1f1e0eb8be3a4bf4d92ce8b/src/net.cpp#L1110-L1117 ?

Reply:
After looking at #16248 and #19191, the current code seems correct.

This line was changed in e5b26deaaa6, which pre-dates the download permission being extracted from noban in #19191: 
```diff 
-    if (addrBind.IsRoutable() && fDiscover && !fWhitelisted)
+    if (addrBind.IsRoutable() && fDiscover && (permissions & PF_NOBAN) == 0)
```
The intent of adding the Download flag in #19191 was: *"It should be possible to grant nodes in a local network (e.g. home, university, enterprise, ...) permission to download blocks even after the maxuploadtarget is hit. Currently this is only possible by setting the noban permission, which has some adverse effects, especially if the peers can't be fully trusted. Fix this by extracting a download permission from noban."*

So, in reply to your questions:

1. IIUC the existing lone bitwise AND here, e.g. `HasAnyFlag(Noban)` is correct

2. I guess we could replace `HasAnyFlag(Noban)` with the more verbose `HasFlag(Noban) || HasFlag(Download)` if that seems clearer

3. Elsewhere, `HasFlag()` appears to be what we want in the cases where it is used, AFAICT

Reply:
Ugh, looks like this is a bug that I introduced in #19191

Reply:
I first thought that too, but here in Bind() I'm not sure.

Reply:
Well, fixing a bug should indeed be done in a separate pull from one that is tagged with [refactoring]

Reply:
Ok will open a separate patch for that.

Reply:
Done in #21644. Resolving here and will simplify this PR, dropping the `HasAnyFlags()` and unit test commits.